File: NeetCode/RM/AH/2.cpp
// https://neetcode.io/problems/is-anagram?list=neetcode150
// Complete

#include <string>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        int s1 = s.size();
        int s2 = t.size();

        unordered_map<char, int> m1;
        unordered_map<char, int> m2;

        if (s1 != s2) {
            return false;
        }

        for (int i = 0; i < s1; i++) {
            char c1 = s[i];
            char c2 = t[i];            
            
            m1[c1]++;
            m2[c2]++;
        }

        for (auto &[key, value]: m1) {
            if (m2[key] != value) {
                return false;
            }
        }
        return true;
    }
};

File: NeetCode/Practice/ML/2.py
# https://neetcode.io/problems/linear-regression-forward

# Ok, not implementing linear regression, just finding it's forward pass

# Complete

import numpy as np
from numpy.typing import NDArray

class Solution:
    
    def get_model_prediction(self, X: NDArray[np.float64], weights: NDArray[np.float64]) -> NDArray[np.float64]:
        # X is an Nx3 NumPy array
        # weights is a 3x1 NumPy array
        # HINT: np.matmul() will be useful

        return np.round(np.matmul(X, weights), 5)


    def get_error(self, model_prediction: NDArray[np.float64], ground_truth: NDArray[np.float64]) -> float:

        return np.round(np.mean(np.square(model_prediction - ground_truth)), 5)
